core/src/saros/concurrent/management/HeartbeatDispatcher.java

```diff
@@ -56,13 +58,15 @@ private void dispatchHeartbeats() {
                     (path, jupiter) -> {
                       JupiterActivity heartbeat =
                           jupiter.generateJupiterActivity(
-                              new NoOperation(), sarosSession.getLocalUser(), path);
+                              new NoOperation(), session.getLocalUser(), path);
                       fireActivity(heartbeat);
                     }));
   }
 
   @Override
   public void stop() {
+    session.removeActivityProducer(this);
     heartbeatScheduledExecutor.shutdown();
+    // FIXME wait here until everything is finished
```

Comment:
Could you add a reason?

Reply:
I thought the reason is obvious. After calling stop I would expect that the component is stopped. If you read the Javadoc and know the the behavior of asyncExec in combination with shutdown it is clear that this component can still do something (not the case anymore after the listener removal). However it is possible that a heartbeat is still broadcasted / executed in the ActivityHandler when you leave the stop method.

Reply:
I will address the FIXME in another patch. (Pushing it later today).